geomesa-features/geomesa-feature-common/src/main/scala/org/locationtech/geomesa/features/ScalaSimpleFeature.scala

```diff
@@ -116,142 +139,180 @@ object ScalaSimpleFeature {
-        reader = null
-        phaser = null
-        true
-      }
-    }
-
-    override lazy val getUserData: java.util.Map[AnyRef, AnyRef] = {
-      val ud = userDataReader.read()
-      userDataReader = null // dereference any backing resources
-      ud
+    /**
+     * Constructor
+     *
+     * @param sft simple feature type
+     * @param id simple feature id
+     * @param reader lazily read attributes, must be thread-safe and already of the appropriate types
+     * @param userDataReader lazily read user data, must be thread-safe
+     *
+     */
+    def this(sft: SimpleFeatureType, id: String, reader: LazyAttributeReader, userDataReader: LazyUserDataReader) = {
+      this(sft)
+      this.id = id
+      this.reader = reader
+      this.userDataReader = userDataReader
     }
 
     override def setAttributeNoConvert(index: Int, value: AnyRef): Unit = {
       if (reader != null) {
         val word = bits.word(index)
         word.synchronized {
           if (word.add(index)) {
-            phaser.arriveAndDeregister()
+            bits.synchronized {
+              if (attributes == null) {
+                attributes = Array.ofDim[AnyRef](sft.getAttributeCount)
+              }
+            }
+            if (count.decrementAndGet() == 0) {
+              // once all attributes have been read, dereference any backing resources so that they can be gc'd
+              reader = null
+              count = null
+            }
           }
         }
       }
       attributes(index) = value
     }
 
+    override def calculateSizeOf(): Long =
+      Sizable.sizeOf(this) + Sizable.deepSizeOf(id, bits, attributes, userData, count, reader, userDataReader)
+  }
+
+  /**
+   * Thread-safe lazy evaluation of attributes
+   */
+  trait LazySimpleFeature extends AbstractSimpleFeature {
+
+    // we synchronize on the low-level words, in order to minimize contention
+    protected val bits: WordBitSet = WordBitSet(getFeatureType.getAttributeCount + 1)
+    protected var attributes: Array[AnyRef] = _
+    protected var userData: java.util.Map[AnyRef, AnyRef] = _
+
+    protected var count = new AtomicInteger(getFeatureType.getAttributeCount)
+    protected var reader: LazyAttributeReader = _
+    protected var userDataReader: LazyUserDataReader = _
+
     override def getAttribute(index: Int): AnyRef = {
       if (reader != null) {
         val word = bits.word(index)
         word.synchronized {
           if (word.add(index)) {
+            bits.synchronized {
+              if (attributes == null) {
+                attributes = Array.ofDim[AnyRef](getFeatureType.getAttributeCount)
+              }
+            }
             attributes(index) = reader.read(index)
-            phaser.arriveAndDeregister()
+            if (count.decrementAndGet() == 0) {
+              // once all attributes have been read, dereference any backing resources so that they can be gc'd
+              reader = null
+              count = null
+            }
```

Comment:
Ok, the first time I read through this...  I thought that after getAttribute was called once, then the next time there'd be an NPE since count is null.  Then I realized that we'd need to zero out 'count', so the 'count+first' time would have an NPE.

But then I realized that's not right.  Since the reader would be null, we'd be ok.

Ok, this next time through, I'm wondering if we only decrement count when we read a new attribute?  If so, then I think I understand it. 

If not, then calling getAttribute(0) 'count' times would remove the reader/count and cause trouble....

Reply:
yeah, it's bounded by the `if(word.add(index))` clause, which is an atomic set addition, so it should only be decremented the first time a given attribute is accessed.